src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/AbstractMultiLifetimeComponent.cs

```diff
@@ -47,38 +59,37 @@ public async Task LoadAsync()
             await LoadCoreAsync();
         }
 
-        private async Task LoadCoreAsync()
+        public async Task LoadCoreAsync()
         {
-            TaskCompletionSource<object> loadedSource = null;
-            IMultiLifetimeInstance instance;
+            JoinableTask initializeAsyncTask;
+
             lock (_lock)
             {
-                if (_instance == null)
+                if (!_instanceTaskSource.Task.IsCompleted)
                 {
-                    _instance = CreateInstance();
-                    loadedSource = _loadedSource;
+                    (T instance, JoinableTask initializeAsyncTask) result = CreateInitializedInstanceAsync();
+                    _instanceTaskSource.SetResult(result);
                 }
 
-                instance = _instance;
+                Assumes.True(_instanceTaskSource.Task.IsCompleted);
+
+                // Should throw TaskCancellatedException if already cancelled in Dispose
+                (_, initializeAsyncTask) = _instanceTaskSource.Task.GetAwaiter().GetResult();
             }
 
-            // While all callers should wait on InitializeAsync, 
-            // only one should complete the completion source
-            await instance.InitializeAsync();
-            loadedSource?.SetResult(null);
+            await initializeAsyncTask;
         }
 
         public Task UnloadAsync()
         {
-            IMultiLifetimeInstance instance = null;
-
+            T instance = null;
             lock (_lock)
             {
-                if (_instance != null)
+                if (_instanceTaskSource.Task.IsCompleted)
                 {
-                    instance = _instance;
-                    _instance = null;
-                    _loadedSource = new TaskCompletionSource<object>();
+                    // Should throw TaskCancellatedException if already cancelled in Dispose
+                    (instance, _) = _instanceTaskSource.Task.GetAwaiter().GetResult();
+                    _instanceTaskSource = new TaskCompletionSource<(T instance, JoinableTask initializeAsyncTask)>();
```

Comment:
Probably want to pass `TaskCreationOptions.RunContinuationsAsynchronously` here to prevent continuations running synchronously when the lock is held and `SetResult` or `TrySetCancelled` is called.